docassemble/AssemblyLine/al_general.py

```diff
@@ -296,6 +366,22 @@ def on_one_line(
             output += ", " + country_name(self._get_country())
         return output
 
+    def normalized_address(self) -> Union[Address, "ALAddress"]:
+        """
+        Try geocoding the address, and if it succeeds, return the "long" normalized version of
+        the address. All methods are still available, such as my_address.normalized_address().block(), etc.,
+        but note that this will be a standard Address object, not an ALAddress object.
+
+        If geocoding fails, return the version of the address as entered by the user instead.
+        """
+        try:
+            self.geocode()
+        except:
+            pass
+        if self.was_geocoded_successfully() and hasattr(self, "norm_long"):
+            return self.norm_long
```

Comment:
What does `norm_long` look like? You mentioned it working for when the user writes a neighborhood instead of a city, but in my experience (particularly living in Dorchester) is that the normalized version will still use the neighborhood (I've never been sure why, I guess to distinguish between otherwise identical addresses in other neighborhoods).

I guess maybe linking to any docs that google might have on how normalization works would help?

Reply:
`norm_long` will be an Address object that looks a lot like the existing Address object.

I ran into this in the MACourts repository, where the Massachusetts town Marlborough was normalized to "Marlboro" for the "short" form and "Marlborough" for the long form.

Some Boston "neighborhoods" will be left alone in my experience--Dorchester, Brighton, etc. are technically neighborhoods but they have unique post offices.

Example:

```json
      "elements": [
        {
          "_class": "docassemble.playground10ALAddress.al_general.ALIndividual", 
          "address": {
            "_class": "docassemble.playground10ALAddress.al_general.ALAddress", 
            "_geocode_response": {
              "address_components": [
                {
                  "long_name": "123", 
                  "short_name": "123", 
                  "types": [
                    "street_number"
                  ]
                }, 
                {
                  "long_name": "Main Street", 
                  "short_name": "Main St", 
                  "types": [
                    "route"
                  ]
                }, 
                {
                  "long_name": "Marlborough", 
                  "short_name": "Marlboro", 
                  "types": [
                    "locality", 
                    "political"
                  ]
                }, 
                {
                  "long_name": "Middlesex County", 
                  "short_name": "Middlesex County", 
                  "types": [
                    "administrative_area_level_2", 
                    "political"
                  ]
                }, 
                {
                  "long_name": "Massachusetts", 
                  "short_name": "MA", 
                  "types": [
                    "administrative_area_level_1", 
                    "political"
                  ]
                }, 
                {
                  "long_name": "United States", 
                  "short_name": "US", 
                  "types": [
                    "country", 
                    "political"
                  ]
                }, 
                {
                  "long_name": "01752", 
                  "short_name": "01752", 
                  "types": [
                    "postal_code"
                  ]
                }
              ], 
              "formatted_address": "123 Main St, Marlborough, MA 01752, USA", 
              "geometry": {
                "location": {
                  "lat": 42.3469235, 
                  "lng": -71.5472462
                }, 
                "location_type": "RANGE_INTERPOLATED", 
                "viewport": {
                  "northeast": {
                    "lat": 42.3482724802915, 
                    "lng": -71.5458972197085
                  }, 
                  "southwest": {
                    "lat": 42.3455745197085, 
                    "lng": -71.5485951802915
                  }
                }
              }, 
              "place_id": "EicxMjMgTWFpbiBTdCwgTWFybGJvcm91Z2gsIE1BIDAxNzUyLCBVU0EiMBIuChQKEgldyO00aIvjiREb2C-ujaNs7RB7KhQKEgnr3eHUaYvjiRFWXzgm3ELELA", 
              "types": [
                "street_address"
              ]
            }, 
            "_geocode_success": true, 
            "_geocoded": true, 
            "address": "123 Main Street", 
            "address_label": {
              "_class": "docassemble.base.util.DALazyTemplate", 
              "instanceName": "users[0].address.address_label"
            }, 
            "administrative_area_level_1": "Massachusetts", 
            "city": "Marlborough", 
            "city_label": {
              "_class": "docassemble.base.util.DALazyTemplate", 
              "instanceName": "users[0].address.city_label"
            }, 
            "city_only": false, 
            "country": "US", 
            "county": "Middlesex County", 
            "geolocate_response": {
              "address_components": [
                {
                  "long_name": "123", 
                  "short_name": "123", 
                  "types": [
                    "street_number"
                  ]
                }, 
                {
                  "long_name": "Main Street", 
                  "short_name": "Main St", 
                  "types": [
                    "route"
                  ]
                }, 
                {
                  "long_name": "Marlborough", 
                  "short_name": "Marlboro", 
                  "types": [
                    "locality", 
                    "political"
                  ]
                }, 
                {
                  "long_name": "Middlesex County", 
                  "short_name": "Middlesex County", 
                  "types": [
                    "administrative_area_level_2", 
                    "political"
                  ]
                }, 
                {
                  "long_name": "Massachusetts", 
                  "short_name": "MA", 
                  "types": [
                    "administrative_area_level_1", 
                    "political"
                  ]
                }, 
                {
                  "long_name": "United States", 
                  "short_name": "US", 
                  "types": [
                    "country", 
                    "political"
                  ]
                }, 
                {
                  "long_name": "01752", 
                  "short_name": "01752", 
                  "types": [
                    "postal_code"
                  ]
                }
              ], 
              "formatted_address": "123 Main St, Marlborough, MA 01752, USA", 
              "geometry": {
                "location": {
                  "lat": 42.3469235, 
                  "lng": -71.5472462
                }, 
                "location_type": "RANGE_INTERPOLATED", 
                "viewport": {
                  "northeast": {
                    "lat": 42.3482724802915, 
                    "lng": -71.5458972197085
                  }, 
                  "southwest": {
                    "lat": 42.3455745197085, 
                    "lng": -71.5485951802915
                  }
                }
              }, 
              "place_id": "EicxMjMgTWFpbiBTdCwgTWFybGJvcm91Z2gsIE1BIDAxNzUyLCBVU0EiMBIuChQKEgldyO00aIvjiREb2C-ujaNs7RB7KhQKEgnr3eHUaYvjiRFWXzgm3ELELA", 
              "types": [
                "street_address"
              ]
            }, 
            "geolocate_success": true, 
            "geolocated": true, 
            "has_no_address": false, 
            "has_no_address_explanation_help": {
              "_class": "docassemble.base.util.DALazyTemplate", 
              "instanceName": "users[0].address.has_no_address_explanation_help"
            }, 
            "has_no_address_explanation_label": {
              "_class": "docassemble.base.util.DALazyTemplate", 
              "instanceName": "users[0].address.has_no_address_explanation_label"
            }, 
            "has_no_address_label": {
              "_class": "docassemble.base.util.DALazyTemplate", 
              "instanceName": "users[0].address.has_no_address_label"
            }, 
            "instanceName": "users[0].address", 
            "locality": "Marlborough", 
            "location": {
              "_class": "docassemble.base.util.LatitudeLongitude", 
              "description": "123 Main Street [NEWLINE] Marlborough, MA 01752", 
              "gathered": true, 
              "instanceName": "users[0].address.location", 
              "known": true, 
              "latitude": 42.3469235, 
              "longitude": -71.5472462
            }, 
            "norm": {
              "_class": "docassemble.playground10ALAddress.al_general.ALAddress", 
              "_geocode_response": {
                "address_components": [
                  {
                    "long_name": "123", 
                    "short_name": "123", 
                    "types": [
                      "street_number"
                    ]
                  }, 
                  {
                    "long_name": "Main Street", 
                    "short_name": "Main St", 
                    "types": [
                      "route"
                    ]
                  }, 
                  {
                    "long_name": "Marlborough", 
                    "short_name": "Marlboro", 
                    "types": [
                      "locality", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "Middlesex County", 
                    "short_name": "Middlesex County", 
                    "types": [
                      "administrative_area_level_2", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "Massachusetts", 
                    "short_name": "MA", 
                    "types": [
                      "administrative_area_level_1", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "United States", 
                    "short_name": "US", 
                    "types": [
                      "country", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "01752", 
                    "short_name": "01752", 
                    "types": [
                      "postal_code"
                    ]
                  }
                ], 
                "formatted_address": "123 Main St, Marlborough, MA 01752, USA", 
                "geometry": {
                  "location": {
                    "lat": 42.3469235, 
                    "lng": -71.5472462
                  }, 
                  "location_type": "RANGE_INTERPOLATED", 
                  "viewport": {
                    "northeast": {
                      "lat": 42.3482724802915, 
                      "lng": -71.5458972197085
                    }, 
                    "southwest": {
                      "lat": 42.3455745197085, 
                      "lng": -71.5485951802915
                    }
                  }
                }, 
                "place_id": "EicxMjMgTWFpbiBTdCwgTWFybGJvcm91Z2gsIE1BIDAxNzUyLCBVU0EiMBIuChQKEgldyO00aIvjiREb2C-ujaNs7RB7KhQKEgnr3eHUaYvjiRFWXzgm3ELELA", 
                "types": [
                  "street_address"
                ]
              }, 
              "_geocoded": true, 
              "address": "123 Main St", 
              "administrative_area_level_1": "MA", 
              "administrative_area_level_2": "Middlesex County", 
              "city": "Marlboro", 
              "city_only": false, 
              "country": "US", 
              "county": "Middlesex County", 
              "geolocate_response": {
                "address_components": [
                  {
                    "long_name": "123", 
                    "short_name": "123", 
                    "types": [
                      "street_number"
                    ]
                  }, 
                  {
                    "long_name": "Main Street", 
                    "short_name": "Main St", 
                    "types": [
                      "route"
                    ]
                  }, 
                  {
                    "long_name": "Marlborough", 
                    "short_name": "Marlboro", 
                    "types": [
                      "locality", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "Middlesex County", 
                    "short_name": "Middlesex County", 
                    "types": [
                      "administrative_area_level_2", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "Massachusetts", 
                    "short_name": "MA", 
                    "types": [
                      "administrative_area_level_1", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "United States", 
                    "short_name": "US", 
                    "types": [
                      "country", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "01752", 
                    "short_name": "01752", 
                    "types": [
                      "postal_code"
                    ]
                  }
                ], 
                "formatted_address": "123 Main St, Marlborough, MA 01752, USA", 
                "geometry": {
                  "location": {
                    "lat": 42.3469235, 
                    "lng": -71.5472462
                  }, 
                  "location_type": "RANGE_INTERPOLATED", 
                  "viewport": {
                    "northeast": {
                      "lat": 42.3482724802915, 
                      "lng": -71.5458972197085
                    }, 
                    "southwest": {
                      "lat": 42.3455745197085, 
                      "lng": -71.5485951802915
                    }
                  }
                }, 
                "place_id": "EicxMjMgTWFpbiBTdCwgTWFybGJvcm91Z2gsIE1BIDAxNzUyLCBVU0EiMBIuChQKEgldyO00aIvjiREb2C-ujaNs7RB7KhQKEgnr3eHUaYvjiRFWXzgm3ELELA", 
                "types": [
                  "street_address"
                ]
              }, 
              "geolocated": true, 
              "instanceName": "users[0].address.norm", 
              "locality": "Marlboro", 
              "location": {
                "_class": "docassemble.base.util.LatitudeLongitude", 
                "description": "123 Main St [NEWLINE] Marlboro, MA 01752", 
                "gathered": true, 
                "instanceName": "users[0].address.norm.location", 
                "known": true, 
                "latitude": 42.3469235, 
                "longitude": -71.5472462
              }, 
              "one_line": "123 Main St, Marlborough, MA 01752, USA", 
              "postal_code": "01752", 
              "route": "Main St", 
              "state": "MA", 
              "street": "Main St", 
              "street_number": "123"
            }, 
            "norm_long": {
              "_class": "docassemble.playground10ALAddress.al_general.ALAddress", 
              "_geocode_response": {
                "address_components": [
                  {
                    "long_name": "123", 
                    "short_name": "123", 
                    "types": [
                      "street_number"
                    ]
                  }, 
                  {
                    "long_name": "Main Street", 
                    "short_name": "Main St", 
                    "types": [
                      "route"
                    ]
                  }, 
                  {
                    "long_name": "Marlborough", 
                    "short_name": "Marlboro", 
                    "types": [
                      "locality", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "Middlesex County", 
                    "short_name": "Middlesex County", 
                    "types": [
                      "administrative_area_level_2", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "Massachusetts", 
                    "short_name": "MA", 
                    "types": [
                      "administrative_area_level_1", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "United States", 
                    "short_name": "US", 
                    "types": [
                      "country", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "01752", 
                    "short_name": "01752", 
                    "types": [
                      "postal_code"
                    ]
                  }
                ], 
                "formatted_address": "123 Main St, Marlborough, MA 01752, USA", 
                "geometry": {
                  "location": {
                    "lat": 42.3469235, 
                    "lng": -71.5472462
                  }, 
                  "location_type": "RANGE_INTERPOLATED", 
                  "viewport": {
                    "northeast": {
                      "lat": 42.3482724802915, 
                      "lng": -71.5458972197085
                    }, 
                    "southwest": {
                      "lat": 42.3455745197085, 
                      "lng": -71.5485951802915
                    }
                  }
                }, 
                "place_id": "EicxMjMgTWFpbiBTdCwgTWFybGJvcm91Z2gsIE1BIDAxNzUyLCBVU0EiMBIuChQKEgldyO00aIvjiREb2C-ujaNs7RB7KhQKEgnr3eHUaYvjiRFWXzgm3ELELA", 
                "types": [
                  "street_address"
                ]
              }, 
              "_geocoded": true, 
              "address": "123 Main Street", 
              "administrative_area_level_1": "Massachusetts", 
              "administrative_area_level_2": "Middlesex County", 
              "city": "Marlborough", 
              "city_only": false, 
              "country": "United States", 
              "county": "Middlesex County", 
              "geolocate_response": {
                "address_components": [
                  {
                    "long_name": "123", 
                    "short_name": "123", 
                    "types": [
                      "street_number"
                    ]
                  }, 
                  {
                    "long_name": "Main Street", 
                    "short_name": "Main St", 
                    "types": [
                      "route"
                    ]
                  }, 
                  {
                    "long_name": "Marlborough", 
                    "short_name": "Marlboro", 
                    "types": [
                      "locality", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "Middlesex County", 
                    "short_name": "Middlesex County", 
                    "types": [
                      "administrative_area_level_2", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "Massachusetts", 
                    "short_name": "MA", 
                    "types": [
                      "administrative_area_level_1", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "United States", 
                    "short_name": "US", 
                    "types": [
                      "country", 
                      "political"
                    ]
                  }, 
                  {
                    "long_name": "01752", 
                    "short_name": "01752", 
                    "types": [
                      "postal_code"
                    ]
                  }
                ], 
                "formatted_address": "123 Main St, Marlborough, MA 01752, USA", 
                "geometry": {
                  "location": {
                    "lat": 42.3469235, 
                    "lng": -71.5472462
                  }, 
                  "location_type": "RANGE_INTERPOLATED", 
                  "viewport": {
                    "northeast": {
                      "lat": 42.3482724802915, 
                      "lng": -71.5458972197085
                    }, 
                    "southwest": {
                      "lat": 42.3455745197085, 
                      "lng": -71.5485951802915
                    }
                  }
                }, 
                "place_id": "EicxMjMgTWFpbiBTdCwgTWFybGJvcm91Z2gsIE1BIDAxNzUyLCBVU0EiMBIuChQKEgldyO00aIvjiREb2C-ujaNs7RB7KhQKEgnr3eHUaYvjiRFWXzgm3ELELA", 
                "types": [
                  "street_address"
                ]
              }, 
              "geolocated": true, 
              "instanceName": "users[0].address.norm_long", 
              "locality": "Marlborough", 
              "location": {
                "_class": "docassemble.base.util.LatitudeLongitude", 
                "description": "123 Main Street [NEWLINE] Marlborough, Massachusetts 01752 [NEWLINE] United States", 
                "gathered": true, 
                "instanceName": "users[0].address.norm_long.location", 
                "known": true, 
                "latitude": 42.3469235, 
                "longitude": -71.5472462
              }, 
              "one_line": "123 Main St, Marlborough, MA 01752, USA", 
              "postal_code": "01752", 
              "route": "Main Street", 
              "state": "Massachusetts", 
              "street": "Main Street", 
              "street_number": "123"
            }, 
            "one_line": "123 Main St, Marlborough, MA 01752, USA", 
            "political": "Marlborough", 
            "postal_code": "01752", 
            "state": "MA", 
            "state_label": {
              "_class": "docassemble.base.util.DALazyTemplate", 
              "instanceName": "users[0].address.state_label"
            }, 
            "street": "Main St", 
            "street_number": "123", 
            "unit": "", 
            "unit_label": {
              "_class": "docassemble.base.util.DALazyTemplate", 
              "instanceName": "users[0].address.unit_label"
            }, 
            "zip": "01752", 
            "zip_label": {
              "_class": "docassemble.base.util.DALazyTemplate", 
              "instanceName": "users[0].address.zip_label"
            }
          }, 
          "instanceName": "users[0]", 
          "location": {
            "_class": "docassemble.base.util.LatitudeLongitude", 
            "gathered": false, 
            "instanceName": "users[0].location", 
            "known": false
          }, 
          "mailing_address": {
            "_class": "docassemble.playground10ALAddress.al_general.ALAddress", 
            "_geocoded": false, 
            "city_only": false, 
            "geolocated": false, 
            "instanceName": "users[0].mailing_address", 
            "location": {
              "_class": "docassemble.base.util.LatitudeLongitude", 
              "gathered": false, 
              "instanceName": "users[0].mailing_address.location", 
              "known": false
            }
          }, 
          "name": {
            "_class": "docassemble.base.util.IndividualName", 
            "instanceName": "users[0].name", 
            "uses_parts": true
          }, 
          "other_addresses": {
            "_class": "docassemble.playground10ALAddress.al_general.ALAddressList", 
            "ask_number": false, 
            "ask_object_type": false, 
            "auto_gather": true, 
            "complete_attribute": null, 
            "elements": [], 
            "instanceName": "users[0].other_addresses", 
            "minimum_number": null, 
            "object_type": {
              "_class": "type", 
              "name": "docassemble.playground10ALAddress.al_general.ALAddress"
            }, 
            "object_type_parameters": {}
          }, 
          "previous_addresses": {
            "_class": "docassemble.playground10ALAddress.al_general.ALAddressList", 
            "ask_number": false, 
            "ask_object_type": false, 
            "auto_gather": true, 
            "complete_attribute": null, 
            "elements": [], 
            "instanceName": "users[0].previous_addresses", 
            "minimum_number": null, 
            "object_type": {
              "_class": "type", 
              "name": "docassemble.playground10ALAddress.al_general.ALAddress"
            }, 
            "object_type_parameters": {}
          }
        }
      ], 
```